incubator/orm/natural_key.go

```diff
@@ -2,101 +2,109 @@ package orm
 
 import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	"github.com/cosmos/cosmos-sdk/types/errors"
 )
 
 var _ Indexable = &NaturalKeyTableBuilder{}
 
-func NewNaturalKeyTableBuilder(prefixData, prefixSeq, prefixIndex byte, key sdk.StoreKey, model NaturalKeyed) *NaturalKeyTableBuilder {
-	if prefixIndex == prefixData || prefixData == prefixSeq {
-		panic("prefixIndex must be unique")
-	}
-
-	builder := NewAutoUInt64TableBuilder(prefixData, prefixSeq, key, model)
-
-	idx := NewUniqueIndex(builder, prefixIndex, func(value interface{}) ([]byte, error) {
-		obj, ok := value.(NaturalKeyed)
-		if !ok {
-			return nil, errors.Wrapf(ErrType, "%T", value)
-		}
-		return obj.NaturalKey(), nil
-	})
+// NewNaturalKeyTableBuilder creates a builder to setup a NaturalKeyTable object.
+func NewNaturalKeyTableBuilder(prefixData byte, storeKey sdk.StoreKey, model NaturalKeyed) *NaturalKeyTableBuilder {
 	return &NaturalKeyTableBuilder{
-		naturalKeyIndex:        idx,
-		AutoUInt64TableBuilder: builder,
+		TableBuilder: NewTableBuilder(prefixData, storeKey, model, Max255DynamicLengthIndexKeyCodec{}),
 	}
 }
 
 type NaturalKeyTableBuilder struct {
-	*AutoUInt64TableBuilder
-	naturalKeyIndex *UniqueIndex
+	*TableBuilder
 }
 
 func (a NaturalKeyTableBuilder) Build() NaturalKeyTable {
-	return NaturalKeyTable{
-		autoTable:       a.AutoUInt64TableBuilder.Build(),
-		naturalKeyIndex: a.naturalKeyIndex,
-	}
+	return NaturalKeyTable{table: a.TableBuilder.Build()}
+
 }
 
 // NaturalKeyed defines an object type that is aware of it's immutable natural key.
 type NaturalKeyed interface {
-	// NaturalKey returns the immutable and serialized natural key of this object
-	NaturalKey() []byte
+	// NaturalKey returns the immutable and serialized natural key of this object. The natural key has to be unique within
+	// it's domain so that not two with same value can exist in the same table.
```

Comment:
I think it's important to have a notice here, and possibly elsewhere in the docs that `255` bytes is the max length for keys. We may want to consider at some point a reverse-encoded VarUInt to get around this limitation. At the same time, I think a primary key in a blockchain application longer than 255 bytes is a problem in and of itself. What do you think?

So for now let's just make sure it's well documented here.

Reply:
I will add some doc to explain the constraints. The `naturalKey()` function is not necessarily limited to 255 char. It is the `Max255DynamicLengthIndexKeyCodec` used with the "default "NaturalKeyTableBuilder". I had moved the constraint into the codec as an extension point to customize the implementation in the future.

Reply:
Okay thanks, will merge now.